crates/sui-bridge-indexer/src/latest_eth_syncer.rs

```diff
@@ -0,0 +1,183 @@
+//! block on Ethereum and will only query for events up to that block number.
+
+use ethers::providers::{Http, Middleware, Provider};
+use ethers::types::Address as EthAddress;
+use mysten_metrics::spawn_logged_monitored_task;
+use std::collections::HashMap;
+use std::sync::Arc;
+use sui_bridge::error::BridgeResult;
+use sui_bridge::eth_client::EthClient;
+use sui_bridge::retry_with_max_elapsed_time;
+use sui_bridge::types::EthLog;
+use tokio::sync::watch;
+use tokio::task::JoinHandle;
+use tokio::time::{self, Duration};
+use tracing::error;
+
+const ETH_LOG_QUERY_MAX_BLOCK_RANGE: u64 = 1000;
+const ETH_EVENTS_CHANNEL_SIZE: usize = 1000;
+const BLOCK_QUERY_INTERVAL: Duration = Duration::from_secs(2);
+
+pub struct LatestEthSyncer<P> {
+    eth_client: Arc<EthClient<P>>,
+    provider: Arc<Provider<Http>>,
+    contract_addresses: EthTargetAddresses,
+}
+
+/// Map from contract address to their start block.
+pub type EthTargetAddresses = HashMap<EthAddress, u64>;
+
+#[allow(clippy::new_without_default)]
+impl<P> LatestEthSyncer<P>
+where
+    P: ethers::providers::JsonRpcClient + 'static,
+{
+    pub fn new(
+        eth_client: Arc<EthClient<P>>,
+        provider: Arc<Provider<Http>>,
+        contract_addresses: EthTargetAddresses,
+    ) -> Self {
+        Self {
+            eth_client,
+            provider,
+            contract_addresses,
+        }
+    }
+
+    pub async fn run(
+        self,
+    ) -> BridgeResult<(
+        Vec<JoinHandle<()>>,
+        mysten_metrics::metered_channel::Receiver<(EthAddress, u64, Vec<EthLog>)>,
+        watch::Receiver<u64>,
+    )> {
+        let (eth_evnets_tx, eth_events_rx) = mysten_metrics::metered_channel::channel(
+            ETH_EVENTS_CHANNEL_SIZE,
+            &mysten_metrics::get_metrics()
+                .unwrap()
+                .channel_inflight
+                .with_label_values(&["eth_events_queue"]),
+        );
+        let latest_block = self.provider.get_block_number().await?.as_u64();
+        let (latest_block_tx, latest_block_rx) = watch::channel(latest_block);
+        let mut task_handles = vec![];
+        task_handles.push(spawn_logged_monitored_task!(Self::run_block_refresh_task(
+            latest_block_tx,
+            self.provider
+        )));
+        for (contract_address, start_block) in self.contract_addresses {
+            let eth_events_tx_clone = eth_evnets_tx.clone();
+            let latest_block_rx_clone = latest_block_rx.clone();
+            let eth_client_clone = self.eth_client.clone();
+            task_handles.push(spawn_logged_monitored_task!(
+                Self::run_event_listening_task(
+                    contract_address,
+                    start_block,
+                    latest_block_rx_clone,
+                    eth_events_tx_clone,
+                    eth_client_clone,
+                )
+            ));
+        }
+        Ok((task_handles, eth_events_rx, latest_block_rx))
+    }
+
+    async fn run_block_refresh_task(
```

Comment:
I think for latest blocks, we don't need this additional channel, we can simply query `get_block_number` in `run_event_listening_task` and only proceed when the latest block number has advanced. This would simplify the code.

`eth_syncer` is made it because originally i planned to add the monitor code in bridge node, so there will be multiple consumers to read the latest block